File: pythonTest/first_non_repeating_character.py
"""
Problem:
Given a string, find the first non-repeating character.
If no such character exists, return None.

Example:
Input: "aabbcdd"
Output: "c"

Approach:
1. Count the frequency of each character using a dictionary.
2. Traverse the string again and return the first character with count == 1.

Why this problem:
- Tests string handling
- Tests dictionary usage
- Very common in coding assessments

Time Complexity: O(n)
Space Complexity: O(n)
"""

def first_non_repeating_char(s):
    # Edge case: empty or None input
    if not s:
        return None

    frequency = {}

    # First pass: count frequency of each character
    for char in s:
        frequency[char] = frequency.get(char, 0) + 1

    # Second pass: find first character that appears once
    for char in s:
        if frequency[char] == 1:
            return char

    # If no non-repeating character exists
    return None


# Simple test cases
if __name__ == "__main__":
    print(first_non_repeating_char("aabbcdd"))  # Expected: c
    print(first_non_repeating_char("aabb"))     # Expected: None
    print(first_non_repeating_char("swiss"))    # Expected: w

File: javaTest/FirstNonRepeatingCharacter.java
import java.util.HashMap;
import java.util.Map;

/*
 Problem:
 Given a string, find the first non-repeating character.
 If no such character exists, return null.

 Example:
 Input: "aabbcdd"
 Output: c

 Approach:
 1. Count the frequency of each character using a HashMap.
 2. Traverse the string again and return the first character with count == 1.

 Time Complexity: O(n)
 Space Complexity: O(n)
*/

public class FirstNonRepeatingCharacter {

    public static Character firstNonRepeatingChar(String s) {
        if (s == null || s.isEmpty()) {
            return null;
        }

        Map<Character, Integer> frequencyMap = new HashMap<>();

        // First pass: count character frequency
        for (char c : s.toCharArray()) {
            frequencyMap.put(c, frequencyMap.getOrDefault(c, 0) + 1);
        }

        // Second pass: find first non-repeating character
        for (char c : s.toCharArray()) {
            if (frequencyMap.get(c) == 1) {
                return c;
            }
        }

        return null; // No non-repeating character found
    }

    // Simple test cases
    public static void main(String[] args) {
        System.out.println(firstNonRepeatingChar("aabbcdd")); // c
        System.out.println(firstNonRepeatingChar("aabb"));    // null
        System.out.println(firstNonRepeatingChar("swiss"));   // w
    }
}
